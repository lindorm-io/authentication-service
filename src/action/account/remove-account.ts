import Joi from "@hapi/joi";
import { IAuthContext } from "../../typing";
import { assertAccountAdmin } from "../../support";

export interface IRemoveAccountOptions {
  accountId: string;
}

const schema = Joi.object({
  accountId: Joi.string().guid().required(),
});

export const removeAccount = (ctx: IAuthContext) => async (options: IRemoveAccountOptions): Promise<void> => {
  await schema.validateAsync(options);

  const { account: admin, logger, repository } = ctx;
  const { accountId } = options;

  const account = await repository.account.find({ id: accountId });

  await assertAccountAdmin(ctx)();

  await repository.account.remove(account);

  logger.debug("account removed", {
    adminId: admin.id,
    accountId: account.id,
  });
};
