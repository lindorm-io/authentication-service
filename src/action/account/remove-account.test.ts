import { removeAccount } from "./remove-account";
import { getTestRepository, getTestAccount, inMemoryStore, logger, resetStore } from "../../test";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("../../support", () => ({
  assertAccountAdmin: jest.fn(() => () => undefined),
}));

describe("removeAccount", () => {
  let ctx: any;

  beforeEach(async () => {
    ctx = {
      account: getTestAccount("email@lindorm.io"),
      logger,
      repository: await getTestRepository(),
    };
    await ctx.repository.account.create(ctx.account);
  });

  afterEach(resetStore);

  test("should remove account", async () => {
    await expect(
      removeAccount(ctx)({
        accountId: "be3a62d1-24a0-401c-96dd-3aff95356811",
      }),
    ).resolves.toBe(undefined);
    expect(inMemoryStore).toMatchSnapshot();
  });
});
