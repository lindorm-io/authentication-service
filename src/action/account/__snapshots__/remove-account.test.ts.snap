// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`removeAccount should remove account 1`] = `
Object {
  "authentication": Object {
    "account": Object {
      "data": Array [],
      "indices": Array [
        Object {
          "index": Object {
            "id": 1,
          },
          "options": Object {
            "unique": true,
          },
        },
      ],
    },
  },
}
`;
